optuna/artifacts/_boto3.py

```diff
@@ -0,0 +1,71 @@
+from __future__ import annotations
+
+import io
+import shutil
+from typing import TYPE_CHECKING
+
+import boto3
+from botocore.exceptions import ClientError
+
+from optuna.artifacts.exceptions import ArtifactNotFound
+
+
+if TYPE_CHECKING:
+    from typing import BinaryIO
+
+    from mypy_boto3_s3 import S3Client
+
+
+class Boto3ArtifactStore:
+    """An artifact backend for Boto3."""
+
+    def __init__(
+        self, bucket_name: str, client: S3Client | None = None, *, avoid_buf_copy: bool = False
+    ) -> None:
+        self.bucket = bucket_name
+        self.client = client or boto3.client("s3")
+        # This flag is added to avoid that upload_fileobj() method of Boto3 client may close the
+        # source file object. See https://github.com/boto/boto3/issues/929.
+        self._avoid_buf_copy = avoid_buf_copy
+
+    def open_reader(self, artifact_id: str) -> BinaryIO:
+        try:
+            obj = self.client.get_object(Bucket=self.bucket, Key=artifact_id)
+        except ClientError as e:
+            if _is_not_found_error(e):
+                raise ArtifactNotFound("not found") from e
```

Comment:
[nit] I've received feedback from Optuna Dashboard users suggesting that it's unclear which artifact wasn't found. It might be out of the scope of this PR, but what do you think about including the artifact_id in the error message?

```suggestion
                raise ArtifactNotFound(f"artifact={artifact_id} is not found") from e
```

Reply:
Fair enough. Is it better to add bucket name as well?
```suggestion
                raise ArtifactNotFound(f"bucket: {self.bucket}, artifact: {artifact_id} was not found") from e
```

Reply:
Sounds great to me 👍

Reply:
I added some more explanation. Please take another look to confirm if they sounds natural.